iommi/sql_trace.py

```diff
@@ -0,0 +1,401 @@
+    elif isinstance(params, bytes):
+        return repr(params.decode(errors="replace"))
+    else:
+        return repr(params)
+
+
+state = threading.local()
+
+
+def set_sql_debug(new_state, validate=True):
+    if new_state == 'None':
+        new_state = None
+    if validate:
+        assert new_state is None or new_state in SQL_DEBUG_LEVELS
+    setattr(state, 'sql_debug', new_state)
+
+
+def get_sql_debug():
+    """ :rtype: bool """
+    result = getattr(state, 'sql_debug', None)
+    if result is not None:
+        return result
+    return getattr(settings, 'SQL_DEBUG', False)
+
+
+@contextmanager
+def no_sql_debug():
+    """
+    Context manager to temporarily suspend sql logging.
+
+    This is useful inside the sql debug implementation to avoid infinite recursion.
+    """
+    old_state = get_sql_debug()
+    set_sql_debug(False)
+    yield
+    set_sql_debug(old_state)
+
+
+SQL_DEBUG_LEVEL_ALL = 'all'
+SQL_DEBUG_LEVEL_ALL_WITH_STACKS = 'stacks'
+SQL_DEBUG_LEVEL_WORST = 'worst'
+
+
+SQL_DEBUG_LEVELS = {
+    SQL_DEBUG_LEVEL_ALL,
+    SQL_DEBUG_LEVEL_ALL_WITH_STACKS,
+    SQL_DEBUG_LEVEL_WORST,
+    False,
+}
+
+settings.SQL_DEBUG = getattr(settings, 'SQL_DEBUG', False)
+
+assert settings.SQL_DEBUG in SQL_DEBUG_LEVELS, f'SQL_DEBUG must be one of: {SQL_DEBUG_LEVELS}'
+
+
+def sql_debug(msg, extra):
+    if get_sql_debug():
+        if 'sql' not in extra:
+            # If we don't do this we'll end up with infinite recursion back here
+            extra['sql'] = ''
+        log.log(level=SQL, msg=msg, extra=extra)
+
+
+def sql_debug_log_to_request(**data):
+    request = get_current_request()
+    if request is not None:
+        try:
+            request.sql_debug_log.append(data)
+        except AttributeError:
+            request.sql_debug_log = [data]
+
+    level = get_sql_debug()
+    if level is False or level == SQL_DEBUG_LEVEL_WORST:
+        return
+    sql_debug_trace_sql(**data)
+
+
+def sql_debug_trace_sql(sql, params=None, **kwargs):
+    if params:
+        sql %= safe_unicode_literal(params)
+    if len(sql) > 10000:
+        sql = '%s... [%d bytes sql]' % (re.sub(r'[\x00-\x08\x0b-\x1f\x80-\xff].*', '.', sql[:10000]), len(sql))
+    else:
+        sql = re.sub(r'[\x00-\x08\x0b-\x1f\x80-\xff]', '.', sql)
+
+    if get_sql_debug() == SQL_DEBUG_LEVEL_ALL_WITH_STACKS:
+        sql_debug(sql_debug_format_stack_trace(kwargs['frame']), extra={})
+
+    kwargs['sql'] = True
+    sql_debug(sql, extra=kwargs)
+    return sql
+
+
+def sql_debug_total_time():
+    try:
+        log = get_current_request().sql_debug_log
+    except AttributeError:
+        return None
+
+    return sum(x['duration'] for x in log)
+
+
+def sql_debug_format_stack_trace(frame):
+    lines = traceback.format_stack(frame)
+
+    base_path = os.path.abspath(os.path.join(settings.BASE_DIR, '..')) + "/"
+    msg = []
+    skip_template_code = False
+    for line in itertools.dropwhile(lambda l: ('/lib/python' in l
```

Comment:
newline before the string

Reply:
That didn't improve it. I moved it to a function instead.